[Flang][OpenMP]Support for lowering task_reduction and in_reduction to MLIR

kaviya2510 opened the pull request:
Added support for lowering of task_reduction and in_reduction to MLIR 
Fixed the below issue which is observed while lowering in_reduction to MLIR.

The below testcase is not generating the expected MLIR.

```fortran
test.f90
subroutine omp_task_in_reduction()
   integer i
   i = 0
   !$omp task in_reduction(+:i)
   i = i + 1
   !$omp end task
end subroutine omp_task_in_reduction
```

It adds some information(mentioned below) related to `omp.private` in the MLIR lowering and this issue is not observed while  enclosing in_reduction inside a parallel construct. 
This issue has also been resolved as part of this PR.

```mlir
  omp.private {type = firstprivate} @_QFomp_task_in_reductionEi_firstprivate_ref_i32 : !fir.ref<i32> alloc {
  ^bb0(%arg0: !fir.ref<i32>):
    %0 = fir.alloca i32 {bindc_name = "i", pinned, uniq_name = "_QFomp_task_in_reductionEi"}
    %1:2 = hlfir.declare %0 {uniq_name = "_QFomp_task_in_reductionEi"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
    omp.yield(%1#0 : !fir.ref<i32>)
  } copy {
  ^bb0(%arg0: !fir.ref<i32>, %arg1: !fir.ref<i32>):
    %0 = fir.load %arg0 : !fir.ref<i32>
    hlfir.assign %0 to %arg1 : i32, !fir.ref<i32>
    omp.yield(%arg1 : !fir.ref<i32>)
  }
```

kiranchandramohan reviewed (commented):
Please add the translation from OpenMP dialect to LLVMIR before lowering to MLIR. Otherwise this will manifest as a crash.

kaviya2510 commented:
> Please add the translation from OpenMP dialect to LLVMIR before lowering to MLIR. Otherwise this will manifest as a crash.

Sure, I will do it.

skatrak reviewed (commented):
Thank you @kaviya2510 for this work. The approach seems fine to me, just adding some comments to hopefully help you rebase this patch whenever it's time.

kaviya2510 committed:
[Flang][OpenMP]Support for lowering task_reduction and in_reduction to MLIR

kaviya2510 commented:
Thanks for the review @skatrak . I rebased and also addressed your comments.
Kindly take a look at it and let me know if it needs any other changes.

skatrak commented on flang/lib/Lower/OpenMP/ClauseProcessor.h:
```diff
@@ -104,6 +104,9 @@ class ClauseProcessor {
       llvm::SmallVectorImpl<const semantics::Symbol *> &isDeviceSyms) const;
+  bool processInReduction(
+      mlir::Location currentLocation, mlir::omp::InReductionClauseOps &result,
+      llvm::SmallVectorImpl<const semantics::Symbol *> &InReductionSyms) const;
```
Nit: I think it's better to specify that it's intended as an output rather than restating the full clause name. Feel free to disagree, but in that case please fix capitalization.
```suggestion
      llvm::SmallVectorImpl<const semantics::Symbol *> &outReductionSyms) const;
```

kaviya2510 replied:
Thanks for the reviews @skatrak. 
yeah, I can understand. I will modify it.

skatrak replied:
Looking at it again I changed my mind on this 😅, so feel free to use whichever way you prefer, as long as the first letter is lowercase.

skatrak commented on flang/lib/Lower/OpenMP/ClauseProcessor.h:
```diff
@@ -122,6 +125,10 @@ class ClauseProcessor {
+  bool processTaskReduction(mlir::Location currentLocation,
+                            mlir::omp::TaskReductionClauseOps &result,
+                            llvm::SmallVectorImpl<const semantics::Symbol *>
+                                &TaskReductionSyms) const;
```
```suggestion
                                &outReductionSyms) const;
```

skatrak commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -344,8 +344,13 @@ void DataSharingProcessor::collectSymbols(
   // Collect all symbols referenced in the evaluation being processed,
   // that matches 'flag'.
   llvm::SetVector<const semantics::Symbol *> allSymbols;
+  bool collectSymbols = true;
+  for (const omp::Clause &clause : clauses) {
+    if (clause.id == llvm::omp::Clause::OMPC_in_reduction)
+      collectSymbols = false;
+  }
```
Nit: This can be simplified with `llvm::find_if()`.

kaviya2510 replied:
Done.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1248,11 +1248,11 @@ static void genTargetEnterExitUpdateDataClauses(
+    lower::AbstractConverter &converter, semantics::SemanticsContext &semaCtx,
+    lower::StatementContext &stmtCtx, const List<Clause> &clauses,
+    mlir::Location loc, mlir::omp::TaskOperands &clauseOps,
+    llvm::SmallVectorImpl<const semantics::Symbol *> &InReductionSyms) {
```
```suggestion
    llvm::SmallVectorImpl<const semantics::Symbol *> &inReductionSyms) {
```

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1261,20 +1261,21 @@ static void genTaskClauses(lower::AbstractConverter &converter,
   cp.processMergeable(clauseOps);
   cp.processPriority(stmtCtx, clauseOps);
   cp.processUntied(clauseOps);
+  cp.processInReduction(loc, clauseOps, InReductionSyms);
```
Nit: Move above `cp.processMergeable()` to keep alphabetical sorting.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1884,7 +1885,9 @@ genTaskOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
   lower::StatementContext stmtCtx;
   mlir::omp::TaskOperands clauseOps;
-  genTaskClauses(converter, semaCtx, stmtCtx, item->clauses, loc, clauseOps);
+  llvm::SmallVector<const semantics::Symbol *> InReductionSyms;
```
```suggestion
  llvm::SmallVector<const semantics::Symbol *> inReductionSyms;
```

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1901,22 +1904,35 @@ genTaskOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+  auto taskOp =
+      genOpWithBody<mlir::omp::TaskOp>(genInfo, queue, item, clauseOps);
+
+  llvm::SmallVector<mlir::Type> inReductionTypes;
+  for (const auto &inreductionVar : clauseOps.inReductionVars)
+    inReductionTypes.push_back(inreductionVar.getType());
+
+  // Add reduction variables as entry block arguments to the task region
+  llvm::SmallVector<mlir::Location> blockArgLocs(InReductionSyms.size(), loc);
+  taskOp->getRegion(0).addArguments(inReductionTypes, blockArgLocs);
```
This isn't necessary. It's done by the call to `genEntryBlock` inside of `genRegionEntryCB`.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1926,13 +1942,26 @@ genTaskgroupOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+  auto taskgroupOp =
+      genOpWithBody<mlir::omp::TaskgroupOp>(genInfo, queue, item, clauseOps);
+
+  llvm::SmallVector<mlir::Type> taskReductionTypes;
+  for (const auto &taskreductionVar : clauseOps.taskReductionVars)
+    taskReductionTypes.push_back(taskreductionVar.getType());
+
+  // Add reduction variables as entry block arguments to the taskgroup region
+  llvm::SmallVector<mlir::Location> blockArgLocs(taskReductionSyms.size(), loc);
+  taskgroupOp->getRegion(0).addArguments(taskReductionTypes, blockArgLocs);
```
Please replace this with an approach similar to what's done in `genTaskOp`. The callback to generate the entry block should be the one that introduces these arguments and binds them to their corresponding symbol. Otherwise, the outside values will be used instead by nested operations.

skatrak reviewed (commented):
Thank you Kaviya, I only have a few small comments.

kaviya2510 committed:
[Flang][OpenMP] Addressed review comments

skatrak commented on flang/lib/Lower/OpenMP/ClauseProcessor.h:
```diff
@@ -105,6 +105,9 @@ class ClauseProcessor {
   bool processIsDevicePtr(
       mlir::omp::IsDevicePtrClauseOps &result,
       llvm::SmallVectorImpl<const semantics::Symbol *> &isDeviceSyms) const;
+  bool processInReduction(
```
Nit: It should go above `processIsDevicePtr`.

skatrak commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -344,8 +344,15 @@ void DataSharingProcessor::collectSymbols(
+
   converter.collectSymbolSet(eval, allSymbols, flag,
-                             /*collectSymbols=*/true,
+                             /*collectSymbols=*/collectSymbols,
```
Nit: The comment is redundant when passing a variable instead of a constant.
```suggestion
                             collectSymbols,
```

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1904,22 +1906,27 @@ genTaskOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
-  return genOpWithBody<mlir::omp::TaskOp>(
+  OpWithBodyGenInfo genInfo =
       OpWithBodyGenInfo(converter, symTable, semaCtx, loc, eval,
                         llvm::omp::Directive::OMPD_task)
           .setClauses(&item->clauses)
           .setDataSharingProcessor(&dsp)
-          .setGenRegionEntryCb(genRegionEntryCB),
-      queue, item, clauseOps);
+          .setGenRegionEntryCb(genRegionEntryCB);
+
+  auto taskOp =
+      genOpWithBody<mlir::omp::TaskOp>(genInfo, queue, item, clauseOps);
+  return taskOp;
```
Nit: It doesn't look like this change is necessary. Since the general convention in this file is to return the result of `genOpWithBody` directly and construct the `OpWithBodyGenInfo` structure parameter inside of the call whenever possible, I think this should be left as it was.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1929,13 +1936,28 @@ genTaskgroupOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+
+  OpWithBodyGenInfo genInfo =
       OpWithBodyGenInfo(converter, symTable, semaCtx, loc, eval,
                         llvm::omp::Directive::OMPD_taskgroup)
-          .setClauses(&item->clauses),
-      queue, item, clauseOps);
+          .setClauses(&item->clauses)
+          .setGenRegionEntryCb(genRegionEntryCB);
+
+  auto taskgroupOp =
+      genOpWithBody<mlir::omp::TaskgroupOp>(genInfo, queue, item, clauseOps);
+  return taskgroupOp;
```
Nit: Same comment as above. For consistency, return directly the result of `genOpWithBody` and construct the `genInfo` argument within the argument list.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1929,13 +1936,28 @@ genTaskgroupOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+  taskgroupArgs.taskReduction.vars = clauseOps.taskReductionVars;
+
+  auto genRegionEntryCB = [&](mlir::Operation *op) {
+    genEntryBlock(converter.getFirOpBuilder(), taskgroupArgs, op->getRegion(0));
+    return llvm::to_vector(taskgroupArgs.getSyms());
```
This is missing the binding of the symbols to the new entry block arguments:
```suggestion
    genEntryBlock(converter.getFirOpBuilder(), taskgroupArgs, op->getRegion(0));
    bindEntryBlockArgs(converter,
                       llvm::cast<mlir::omp::BlockArgOpenMPOpInterface>(op),
                       taskgroupArgs);
    return llvm::to_vector(taskgroupArgs.getSyms());
```

skatrak commented on flang/test/Lower/OpenMP/taskgroup-task-array-reduction.f90:
```diff
@@ -0,0 +1,49 @@
+! CHECK:            %[[VAL_3:.*]] = fir.alloca !fir.box<!fir.array<?xf32>>
+! CHECK:            fir.store %[[VAL_2]]#1 to %[[VAL_3]] : !fir.ref<!fir.box<!fir.array<?xf32>>>
+! CHECK:            omp.taskgroup task_reduction(byref @add_reduction_byref_box_Uxf32 %[[VAL_3]] ->  %[[VAL_4:.*]]: !fir.ref<!fir.box<!fir.array<?xf32>>>) {
+! CHECK:              %[[VAL_5:.*]] = fir.alloca !fir.box<!fir.array<?xf32>>
```
This doesn't look right to me, but I might be wrong. Shouldn't we be passing the result of an `hlfir.declare` of VAL_4 to `omp.task`, rather than creating an alloca for it? Perhaps this is related to the missing binding of Fortran symbols to entry block arguments during the creation of the region for `omp.taskgroup`.

skatrak commented on flang/test/Lower/OpenMP/taskgroup-task_reduction01.f90:
```diff
@@ -0,0 +1,34 @@
+!CHECK:         %[[VAL_0:.*]] = fir.alloca i32 {bindc_name = "res", uniq_name = "_QFomp_taskgroup_task_reductionEres"}
+!CHECK:         %[[VAL_1:.*]]:2 = hlfir.declare %[[VAL_0]] {uniq_name = "_QFomp_taskgroup_task_reductionEres"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
+!CHECK:         omp.taskgroup task_reduction(@[[RED_I32_NAME]]  %[[VAL_1]]#0 -> %[[VAL_2:.*]] : !fir.ref<i32>) {
+!CHECK:           %[[VAL_3:.*]] = fir.load %[[VAL_1]]#0 : !fir.ref<i32>
```
The `omp.taskgroup` region shouldn't have references to VAL_1. References to that variable should be done through VAL_2.

skatrak commented on flang/test/Lower/OpenMP/taskgroup-task_reduction02.f90:
```diff
@@ -0,0 +1,36 @@
+!CHECK-LABEL:  func.func @_QPin_reduction() {
+!                [...]
+!CHECK:          omp.taskgroup task_reduction(@[[RED_I32_NAME]] %[[VAL_1:.*]]#0 -> %[[VAL_3:.*]] : !fir.ref<i32>) {
+!CHECK:          omp.task in_reduction(@[[RED_I32_NAME]] %[[VAL_1]]#0 -> %[[VAL_4:.*]] : !fir.ref<i32>) {
```
Same problem here: VAL_1 shouldn't be used inside of the `omp.taskgroup` region.

skatrak commented on flang/test/Lower/OpenMP/taskgroup-task-array-reduction.f90:
```diff
@@ -0,0 +1,49 @@
+! CHECK:             return
+! CHECK:           }
+
+subroutine taskReduction(x)
```
Nit: For consistency with other tests.
```suggestion
subroutine task_reduction(x)
```

skatrak reviewed (commented):
Sorry for the delay getting back to this, I hope I haven't blocked you too much. I think there is one minor problem at the moment, which will probably make you have to update some of the tests, but other than that I just have some small nits.

kaviya2510 commented:
@skatrak It's fine. This PR is dependent on the translation patches for inreduction and task_reduction being merged first, so no problem.
Thank you for the feedback again. I’ll review your comments and get back to you soon.

kaviya2510 committed:
Merge branch 'main' into task_reduction

kaviya2510 committed:
Addressed review comments: Binding Fortran symbols to entry block arguments in taskgroup construct and fixed testcases

kaviya2510 commented:
@skatrak apologies for the delayed response. I thought that this patch depends on the LLVM translation patch for task reduction and in-reduction and it might crash without the translation part. So I decided to keep it on hold and waiting for the translation patch to merge first.

Later, I came to know that you had added a patch which throws an error for unhandled clauses during LLVM translation. Realizing that this patch does not have a dependency, I resumed working on it.

kaviya2510 commented:
I have addressed all your review comments in the recent patch. Could you please take a look at it and let me know if you have any comments?
Thankyou.

tblah commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2498,13 +2499,24 @@ genTaskOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+  taskArgs.inReduction.syms = inReductionSyms;
+  taskArgs.inReduction.vars = clauseOps.inReductionVars;
+
+  auto genRegionEntryCB = [&](mlir::Operation *op) {
+    genEntryBlock(converter.getFirOpBuilder(), taskArgs, op->getRegion(0));
+    bindEntryBlockArgs(converter,
+                       llvm::cast<mlir::omp::BlockArgOpenMPOpInterface>(op),
+                       taskArgs);
+    return llvm::to_vector(taskArgs.getSyms());
+  };
```
See https://github.com/llvm/llvm-project/pull/132036

kaviya2510 replied:
Thank you for pointing it out. I have made the necessary changes, could you please take a look at it?

tblah commented on flang/lib/Lower/OpenMP/ReductionProcessor.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "flang/Parser/tools.h"
 #include "mlir/Dialect/OpenMP/OpenMPDialect.h"
 #include "llvm/Support/CommandLine.h"
+#include <type_traits>
```
What is this used for?

kaviya2510 replied:
I am using the template `std::is_same_v` in function `processReductionArguments()`  and included corresponding header  `type_traits`

tblah reviewed (commented):
Thanks for your work on this. Looks pretty good overall. Just some minor points.

kaviya2510 committed:
Addressed review comment: Replacing the call setGenRegionEntryCb() with setEntryBlockArgs()

kaviya2510 commented:
Thanks for the review @tblah.

tblah reviewed (approved):
LGTM, thanks!

kaviya2510 commented:
Hi @skatrak, Could you please review the patch and provide your approval?

kaviya2510 commented:
> LGTM, thanks!

Thank you for the approval @tblah

skatrak commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -988,6 +988,30 @@ bool ClauseProcessor::processIsDevicePtr(
       });
 }
 
+bool ClauseProcessor::processInReduction(
```
Nit: It should go above `ClauseProcessor::processIsDevicePtr` here too.

skatrak reviewed (approved):
Thank you Kaviya, LGTM! I have a minimal nit, but no need for another review by me before merging.

kaviya2510 committed:
Fixed dsa of variables in task construct in the presence of in_reduction clause

kaviya2510 commented:
`According to the OpenMP specification, the rules for variables with implicitly determined data-sharing attributes are:
* In a _parallel_ construct, if no _default_ clause is present, these variables are **shared**
* In a task generating construct, if no _default_ clause is present, a variable for which the data-sharing attribute is not determined by the rules aboveand that in the enclosing context is determined to be shared by all implicit tasks bound to the current team is **shared**
* In an orphaned task generating construct, if no _default_ clause is present, dummy arguments are **firstprivate**
* In a task generating construct, if no _default_ clause is present, a variable for which the data-sharing attribute is not determined by the rules above is **firstprivate**`

```fortran
subroutine omp_task_in_reduction()
   integer i
   i = 0
   !$omp task in_reduction(+:i)
   i = i + 1
   !$omp end task
end subroutine omp_task_in_reduction
```
In the above example, without any explicit data sharing attribute the variable `i` would normally considered as `firstprivate` in task construct. However, since the variable`i` appears in the `in_reduction` clause, its data-sharing attribute should not be firstprivate.

With the current flow, the compiler incorrectly marks the data-sharing attribute of `i` as `firstprivate`, based on the implicit rule for task constructs and as a result it generate the below mlir during lowering.
```mlir
 omp.private {type = firstprivate} @_QFomp_task_in_reductionEi_firstprivate_ref_i32 : !fir.ref<i32> alloc {
  ^bb0(%arg0: !fir.ref<i32>):
    %0 = fir.alloca i32 {bindc_name = "i", pinned, uniq_name = "_QFomp_task_in_reductionEi"}
    %1:2 = hlfir.declare %0 {uniq_name = "_QFomp_task_in_reductionEi"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
    omp.yield(%1#0 : !fir.ref<i32>)
  } copy {
  ^bb0(%arg0: !fir.ref<i32>, %arg1: !fir.ref<i32>):
    %0 = fir.load %arg0 : !fir.ref<i32>
    hlfir.assign %0 to %arg1 : i32, !fir.ref<i32>
    omp.yield(%arg1 : !fir.ref<i32>)
  }
```
I am felling that my earlier fix in DataSharingProcessor.cpp is not correct. Instead of skipping collecting symbols in DataSharingProcessor.cpp, I added a new change which detects the dsa of variable `i` as in_reduction and skip marking it as  firstprivate.

@tblah , could you please take a look at it and review my changes?

tblah commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -409,8 +409,8 @@ void DataSharingProcessor::collectSymbols(
   llvm::SetVector<const semantics::Symbol *> allSymbols;
-  converter.collectSymbolSet(eval, allSymbols, flag,
-                             /*collectSymbols=*/true,
+
+  converter.collectSymbolSet(eval, allSymbols, flag, /*collectSymbols=*/true,
```
ultra-nit: now that this isn't changing there's no need for this file to have been modified at all in this PR. This keeps the git history clearer (e.g. if somebody wanted a `git blame` of this line).

kaviya2510 replied:
Done with the changes.

tblah reviewed (approved):
The new approach sounds good. Sorry I didn't catch this on the last round of review.

kaviya2510 committed:
Fix code formatting

kaviya2510 commented:
Thanks for the review.

kaviya2510 commented:
> Thank you Kaviya, LGTM! I have a minimal nit, but no need for another review by me before merging.

Thanks for the review @skatrak
